### PR TITLE
[WIP] Fix scheduled jobs for branch-3.4 & branch-3.5

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to fix `PySpark Test scheduled jobs` for `branch-3.4` & `branch-3.5`.

### Why are the changes needed?
Fix github jobs.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Thank you, @panbingkun . Please let me know when this PR is ready.

**Comment by dongjoon-hyun** [comment]
I backported #42897 to branch-3.5 and branch-3.4. Could you rebase this PR to the `master` branch once more, @panbingkun .

**Comment by panbingkun** [comment]
> I backported #42897 to branch-3.5 and branch-3.4. Could you rebase this PR to the `master` branch once more, @panbingkun .

Okay, I will verify it first today.

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -801,53 +803,53 @@ jobs:
...
-        java-version: ${{ matrix.java }}
-    - name: Build with Maven
-      run: |
-        export MAVEN_OPTS="-Xss64m -Xmx2g -XX:ReservedCodeCacheSize=1g -Dorg.slf4j.simpleLogger.defaultLogLevel=WARN"
-        export MAVEN_CLI_OPTS="--no-transfer-progress"
-        export JAVA_VERSION=${{ matrix.java }}
-        # It uses Maven's 'install' intentionally, see https://github.com/apache/spark/pull/26414.
-        ./build/mvn $MAVEN_CLI_OPTS -DskipTests -Pyarn -Pkubernetes -Pvolcano -Phive -Phive-thriftserver -Phadoop-cloud -Djava.version=${JAVA_VERSION/-ea} install
-        rm -rf ~/.m2/repository/org/apache/spark
+      - name: Checkout Spark repository
```
Comment (dongjoon-hyun) [inline .github/workflows/build_and_test.yml]: This new indentation looks like a mistake.

Comment (dongjoon-hyun) [reply .github/workflows/build_and_test.yml]: NVM. This is an outdated comment about the previous commit.

Comment (panbingkun) [reply .github/workflows/build_and_test.yml]: Yeah, I'm trying to find a way to verify it faster.

**Comment by panbingkun** [comment]
I temporarily removed some unrelated tests to make it faster.

**Comment by panbingkun** [comment]
Finally, I can validate the logic of the `branch` in the pr submitted by `master`.  Currently, I am `pinning` some python dependency library versions and `verifying` them.

**Comment by panbingkun** [comment]
```
Warning, treated as error:
/__w/spark/spark/python/docs/source/reference/api/pyspark.ml.Estimator.rst:60:autosummary: failed to import Estimator.uid.
Possible hints:
* AttributeError: type object 'Estimator' has no attribute 'uid'
* ModuleNotFoundError: No module named 'Estimator'
* ImportError: 
* AttributeError: type object 'Estimator' has no attribute 'Estimator'
* KeyError: 'Estimator'
* ModuleNotFoundError: No module named 'pyspark.ml.Estimator'
make: *** [Makefile:35: html] Error 2
```

**Comment by panbingkun** [comment]
So far, the tests related to the `PySpark` in `branch-3.5` have `passed`, and the issue of `docs build` is currently being resolved.

**Comment by panbingkun** [comment]
Before the pr https://github.com/apache/spark/pull/44012 was `backport` to `branch-3.5` or `branch-3.4`, there was a issue when sphinx versions >= `3.1.0`.
But in the file `build_and_test.yml` of our `master`, we set `sphinx==4.5.0`
https://github.com/apache/spark/blob/acf17fd67217941891fde0851b351b99e55d7c4a/.github/workflows/build_and_test.yml#L756

File: .github/workflows/build_and_test.yml [hunk 2]
```diff
@@ -341,7 +176,9 @@ jobs:
       fail-fast: false
       matrix:
         java:
-          - ${{ inputs.java }}
+          - 8
```
Comment (panbingkun) [inline .github/workflows/build_and_test.yml:179]: only for test branch-3.5

File: .github/workflows/build_and_test.yml [hunk 3]
```diff
@@ -365,7 +202,7 @@ jobs:
             pyspark-pandas-connect-part3
     env:
       MODULES_TO_TEST: ${{ matrix.modules }}
-      PYTHON_TO_TEST: 'python3.9'
+      PYTHON_TO_TEST: ''
```
Comment (panbingkun) [inline .github/workflows/build_and_test.yml:205]: Same as the value specified in `branch-3.5.yml`

File: .github/workflows/build_and_test.yml [hunk 4]
```diff
@@ -382,17 +219,10 @@ jobs:
       with:
         fetch-depth: 0
         repository: apache/spark
-        ref: ${{ inputs.branch }}
+        ref: ${{ matrix.branch }}
```
Comment (panbingkun) [inline .github/workflows/build_and_test.yml:222]: Only for test

File: .github/workflows/build_and_test.yml [hunk 5]
```diff
@@ -438,11 +268,21 @@ jobs:
         curl -s https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh > miniconda.sh
         bash miniconda.sh -b -p $HOME/miniconda
         rm miniconda.sh
+    - name: Install Python test dependencies for branch-3.4
```
Comment (panbingkun) [inline .github/workflows/build_and_test.yml:271]: In order to `pass` the pyspark testing of `branch-3.4` and `branch-3.5`.

File: .github/workflows/build_and_test.yml [hunk 6]
```diff
@@ -438,11 +268,21 @@ jobs:
...
+      if: matrix.branch == 'branch-3.5'
+      run: |
+        python3.9 -m pip install 'numpy==1.25.1' 'pandas<=2.0.3' 'pyarrow==12.0.1' 'matplotlib==3.7.2' 'torch==2.0.1' 'torchvision==0.15.2' 'scikit-learn==1.1.*'
     # Run the tests.
     - name: Run tests
       env: ${{ fromJSON(inputs.envs) }}
       shell: 'script -q -e -c "bash {0}"'
       run: |
+        export SCALA_PROFILE="scala2.13"
+        unset GITHUB_ACTIONS
```
Comment (panbingkun) [inline .github/workflows/build_and_test.yml:285]: hack git compare

File: .github/workflows/build_and_test.yml [hunk 7]
```diff
@@ -684,32 +438,18 @@ jobs:
...
         python3.9 -m pip install 'pandas-stubs==1.2.0.53' ipython 'grpcio==1.48.1' 'grpc-stubs==1.24.11' 'googleapis-common-protos-stubs==2.2.0'
     - name: Install Python linter dependencies for branch-3.5
-      if: inputs.branch == 'branch-3.5'
       run: |
         # SPARK-45212: Copy from https://github.com/apache/spark/blob/555c8def51e5951c7bf5165a332795e9e330ec9d/.github/workflows/build_and_test.yml#L631-L638
         # Should delete this section after SPARK 3.5 EOL.
-        python3.9 -m pip install 'flake8==3.9.0' pydata_sphinx_theme 'mypy==0.982' 'pytest==7.1.3' 'pytest-mypy-plugins==1.9.3' numpydoc 'jinja2<3.0.0' 'black==22.6.0'
+        python3.9 -m pip install 'flake8==3.9.0' pydata_sphinx_theme 'mypy==0.982' 'pytest==7.1.3' 'pytest-mypy-plugins==1.9.3' 'numpy==1.25.1' 'pyarrow==12.0.1' numpydoc 'jinja2<3.0.0' 'black==22.6.0' 'pandas<=2.0.3' 'matplotlib==3.7.2' 'torch==2.0.1' 'torchvision==0.15.2'
```
Comment (panbingkun) [inline .github/workflows/build_and_test.yml:447]: For fixing as 
```
/usr/local/lib/python3.9/dist-packages/torch/_dynamo/mutation_guard.py:1: error: disable_error_code: Invalid error code(s): method-assign  [misc]
/usr/local/lib/python3.9/dist-packages/torch/_dynamo/eval_frame.py:1: error: disable_error_code: Invalid error code(s): method-assign  [misc]
/usr/local/lib/python3.9/dist-packages/torch/_dynamo/debug_utils.py:1: error: disable_error_code: Invalid error code(s): method-assign  [misc]
python/pyspark/pandas/plot/matplotlib.py:23: error: Module "matplotlib.axes._base" has no attribute "_process_plot_format"  [attr-defined]
Found 4 errors in 4 files (checked 688 source files)
```

**Comment by panbingkun** [comment]
This PR is basically successful, and I will slightly organize it and submit it as a separate PR. This PR allows me to keep it as a reference for future testing of the branch.

**Pull request closed without merging**